kernel/busy_wait.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/timer/system_timer.h>
+#include <zephyr/sys_clock.h>
+#include <kernel_arch_interface.h>
+
+void z_impl_k_busy_wait(uint32_t usec_to_wait)
+{
+	SYS_PORT_TRACING_FUNC_ENTER(k_thread, busy_wait, usec_to_wait);
+	if (usec_to_wait == 0U) {
+		SYS_PORT_TRACING_FUNC_EXIT(k_thread, busy_wait, usec_to_wait);
+		return;
+	}
+
+#if defined(CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT)
+	arch_busy_wait(usec_to_wait);
+#elif defined(CONFIG_SYS_CLOCK_EXISTS)
+	uint32_t start_cycles = k_cycle_get_32();
+
+	/* use 64-bit math to prevent overflow when multiplying */
+	uint32_t cycles_to_wait = (uint32_t)(
+		(uint64_t)usec_to_wait *
+		(uint64_t)sys_clock_hw_cycles_per_sec() /
+		(uint64_t)USEC_PER_SEC
+	);
+
+	for (;;) {
+		uint32_t current_cycles = k_cycle_get_32();
+
+		/* this handles the rollover on an unsigned 32-bit value */
+		if ((current_cycles - start_cycles) >= cycles_to_wait) {
+			break;
+		}
+	}
+#else
+	/*
+	 * Crud busy loop assuming 1 GHz CPU and 2 cycles per loop.
+	 * Reality is certainly much worse but all we want here is a
+	 * ball-park figure that ought to be good enough for the purpose
+	 * of being able to configure out timer support.
+	 * Those metrics could eventually be obtained from DT if needed.
+	 */
+	unsigned int cpu_hz = 1000000000;
+	unsigned int loops_per_usec = cpu_hz / USEC_PER_SEC * 2;
+	unsigned int loops = loops_per_usec * usec_to_wait;
+
+	do {
+		arch_nop();
+	} while (loops-- > 0);
```

Comment:
Two items related to this loop.

1. The combination of both the do-while loop and the post-decrement on _loops_ means that we are likely to go through the loop one more time than intended. I realise that this is a meant to be a crude approximation, so it may not be the most significant of observations.
2. Is it worthwhile to just have a Kconfig value to indicate the number of loops per microsecond? This could serve as a convenient manner to tune this delay until the DT method you described is implemented.

Reply:
1) The timing won't be accurate enough to matter, but I fixed the logic anyway.
2) Good idea, implemented.